src/nal.rs

```diff
@@ -0,0 +1,78 @@
+use crate::{Error, TcpSocket, W5500};
+use embedded_hal::blocking::spi::{Transfer, Write};
+use embedded_hal::digital::v2::OutputPin;
+
+pub struct Interface<CS: OutputPin, SPI: Transfer<u8> + Write<u8>> {
+    w5500: core::cell::RefCell<W5500<CS, SPI>>,
+}
+
+impl<CSE, SPIE, CS, SPI> Interface<CS, SPI>
+where
+    SPI: Transfer<u8, Error = SPIE> + Write<u8, Error = SPIE>,
+    CS: OutputPin<Error = CSE>,
+{
+    pub fn new(w5500: W5500<CS, SPI>) -> Self {
+        Self {
+            w5500: core::cell::RefCell::new(w5500),
+        }
+    }
+}
+
+impl<CSE, SPIE, CS, SPI> embedded_nal::TcpStack for Interface<CS, SPI>
+where
+    SPI: Transfer<u8, Error = SPIE> + Write<u8, Error = SPIE>,
+    CS: OutputPin<Error = CSE>,
+    CSE: core::fmt::Debug,
+    SPIE: core::fmt::Debug,
+{
+    type TcpSocket = TcpSocket;
+    type Error = Error<SPIE, CSE>;
+
+    fn open(&self, _mode: embedded_nal::Mode) -> Result<Self::TcpSocket, Self::Error> {
+        let socket = self.w5500.borrow_mut().open_tcp()?;
+        Ok(socket)
+    }
+
+    fn connect(
```

Comment:
Because this function consumes the socket, we are actually losing the handle to the socket in the case of errors. We should always return the socket when a failure occurs so that the socket can be re-used in the future.